framework/include/actions/MaterialOutputAction.h

```diff
@@ -23,16 +22,35 @@ class MaterialBase;
 class MaterialOutputAction : public Action
 {
 public:
-  /**
-   * Class constructor
-   * @param params Input parameters for this action object
-   */
   static InputParameters validParams();
 
   MaterialOutputAction(const InputParameters & params);
+
   virtual void act() override;
 
+  /// Meta data describing the setup of an output object
+  struct OutputMetaData
+  {
+    std::string _aux_kernel_name;
+    std::string _index_symbols;
+    std::vector<std::string> _param_names;
+  };
+
 protected:
+  /**
+   * A function to be overriden by derived actions to handle a set of material property types
+   */
+  virtual std::vector<std::string> materialOutput(const std::string & property_name,
+                                                  const MaterialBase & material,
+                                                  bool get_names_only);
+
+  /// Universal output object setup function
+  std::vector<std::string> outputHelper(const OutputMetaData & metadata,
+                                        const MaterialPropertyName & property_name,
```

Comment:
```suggestion
                                        const std::string & property_name,
```
functors can be things other than material properties so might as well keep the most generic string type and not implicitly convert it. I know there's still some logical inconsistency with the variable name ... but there's also logical inconsistency with having functors be output here ... but that's not your fault

Reply:
I can use a std::conditional here

Reply:
oh, wait, I cannot (I was thinking about the auxkernel)

Reply:
I added the std::conditional, so that the parameter type shows up correctly in the docs.